bionic/persistence.py

```diff
@@ -1037,113 +1005,49 @@ def from_computation(
-            code_version_minor=code_fingerprint.version.minor,
-        )
-        exact_code_dict = dict(
-            nominal=nominal_code_dict,
-        )
-
-        bytecode_hash = code_fingerprint.bytecode_hash
-        exact_code_dict["bytecode_hash"] = bytecode_hash
-        if treat_bytecode_as_functional:
-            functional_code_dict["bytecode_hash"] = bytecode_hash
-
-        # If the function's output can change with each run, we add the flow uuid to the
-        # hash so that it will be different each time.
-        if can_functionally_change_per_run:
-            functional_code_dict["flow_instance_uuid"] = flow_instance_uuid
-
-        # This section is a bit repetitive, but I'm not sure there's much to be done
-        # about it.
-        functional_deps_list = [
-            dict(
-                descriptor=task_key.dnode.to_descriptor(),
-                hash=provenance_digest.functional_hash,
-            )
-            for task_key, provenance_digest in dep_task_key_provenance_digest_pairs
-        ]
-        nominal_deps_list = [
-            dict(
-                descriptor=task_key.dnode.to_descriptor(),
-                hash=provenance_digest.nominal_hash,
-            )
-            for task_key, provenance_digest in dep_task_key_provenance_digest_pairs
+        sorted_dep_digests = [
+            dep_digest
+            for _, dep_digest in sorted(dep_provenance_digests_by_task_key.items())
         ]
-        exact_deps_list = [
-            dict(
-                descriptor=task_key.dnode.to_descriptor(),
-                hash=provenance_digest.exact_hash,
-            )
-            for task_key, provenance_digest in dep_task_key_provenance_digest_pairs
-        ]
-
         functional_hash = hash_simple_obj_to_hex(
-            dict(
-                code=functional_code_dict,
-                deps=functional_deps_list,
-            )
+            [
+                [dep_digest.functional_hash for dep_digest in sorted_dep_digests],
+                code_fingerprint.orig_flow_name,
+                code_fingerprint.version.major,
+                CACHE_SCHEMA_VERSION,
+                (
+                    code_fingerprint.bytecode_hash
+                    if treat_bytecode_as_functional
+                    else None
+                ),
+                (flow_instance_uuid if can_functionally_change_per_run else None),
+            ]
         )
         nominal_hash = hash_simple_obj_to_hex(
-            dict(
-                code=nominal_code_dict,
-                deps=nominal_deps_list,
-            )
+            [
+                functional_hash,
+                [dep_digest.nominal_hash for dep_digest in sorted_dep_digests],
+                code_fingerprint.version.minor,
+            ]
         )
         exact_hash = hash_simple_obj_to_hex(
-            dict(
-                code=exact_code_dict,
-                deps=exact_deps_list,
-            )
+            [
+                nominal_hash,
```

Comment:
Should this be `functional_hash`?

Reply:
The goal is for each hash to include all the information from the previous hashes -- so if `nominal_hash` changes, we want `exact_hash` to change as well. I can add a comment to make this more explicit.

Reply:
never mind, misread this.